File: src/context.ts
import type { KanbanStore } from "./store.js";
import type { Checkpoint, ContextPacket, Handoff, Task, TaskNote } from "./types.js";

function iso(epochMs: number | null): string {
  return epochMs === null ? "-" : new Date(epochMs).toISOString();
}

function taskLine(task: Task): string {
  return `- ${task.id} [${task.status}] P${task.priority} ${task.title}`;
}

function renderCheckpoint(checkpoint: Checkpoint): string {
  const lines = [
    `### Checkpoint ${checkpoint.seq} · ${checkpoint.state} · ${checkpoint.author} · ${iso(checkpoint.createdAt)}`,
    `Summary: ${checkpoint.summary}`,
    `Intent: ${checkpoint.intent}`,
    `Next action: ${checkpoint.nextAction}`,
  ];
  if (checkpoint.blockers.length) lines.push(`Blockers: ${checkpoint.blockers.join("; ")}`);
  if (checkpoint.validations.length) {
    lines.push(`Validations: ${checkpoint.validations.join("; ")}`);
  }
  const repo = [checkpoint.repoPath, checkpoint.branch, checkpoint.headSha]
    .filter(Boolean)
    .join(" · ");
  if (repo) lines.push(`Repository: ${repo}`);
  if (checkpoint.dirtySummary) lines.push(`Working tree: ${checkpoint.dirtySummary}`);
  return lines.join("\n");
}

function renderNote(note: TaskNote): string {
  return `- ${iso(note.createdAt)} · ${note.kind} · ${note.author}: ${note.body}`;
}

function renderHandoff(handoff: Handoff): string {
  const target = handoff.toAgent ?? "next compatible agent";
  const acceptance = handoff.acceptedBy
    ? `Accepted by: ${handoff.acceptedBy} · ${iso(handoff.acceptedAt)}`
    : "Accepted by: (pending)";
  return [
    `### Handoff ${handoff.id} · ${handoff.status} · ${handoff.reason}`,
    `From: ${handoff.fromAgent} · To: ${target}`,
    `Summary: ${handoff.summary}`,
    `Intent: ${handoff.intent}`,
    `Next action: ${handoff.nextAction}`,
    acceptance,
  ].join("\n");
}

function clip(value: string, max: number): string {
  if (value.length <= max) return value;
  return `${value.slice(0, Math.max(0, max - 15))}…[truncated]`;
}

export function contextPacket(
  store: KanbanStore,
  taskID: string,
  limits: { notes?: number; checkpoints?: number; handoffs?: number } = {},
): ContextPacket {
  return {
    task: store.requireTask(taskID),
    ancestors: store.ancestors(taskID),
    dependencies: store.dependencies(taskID),
    claim: store.getClaim(taskID),
    notes: store.notes(taskID, limits.notes ?? 50),
    checkpoints: store.checkpoints(taskID, limits.checkpoints ?? 10),
    handoffs: store.handoffs({ taskID, limit: limits.handoffs ?? 10 }).reverse(),
    generatedAt: Date.now(),
    truncated: false,
  };
}

// Cold-start context is bounded by dropping oldest history first. Task identity,
// spec, current lease, dependency state, and the newest checkpoint are never
// dropped: those are the minimum safe handoff contract.
export function renderContext(packet: ContextPacket, maxChars = 20_000): string {
  if (!Number.isInteger(maxChars) || maxChars < 1_000) {
    throw new Error("maxChars must be an integer >= 1000");
  }

  const task = packet.task;
  const fixed = [
    "# Kanban cold-start context",
    "",
    `Generated: ${iso(packet.generatedAt)}`,
    "",
    "## Current task",
    `ID: ${task.id}`,
    `Type: ${task.type}`,
    `Status: ${task.status}`,
    `Priority: ${task.priority}`,
    `Title: ${task.title}`,
    `Body: ${task.body ?? "(none)"}`,
    "",
    "## Claim",
    packet.claim
      ? `${packet.claim.agentID} · session ${packet.claim.sessionID ?? "-"} · expires ${iso(packet.claim.expiresAt)}`
      : "unclaimed",
    "",
    "## Ancestry",
    ...(packet.ancestors.length ? packet.ancestors.map(taskLine) : ["(none)"]),
    "",
    "## Dependencies",
    ...(packet.dependencies.length ? packet.dependencies.map(taskLine) : ["(none)"]),
  ].join("\n");

  const newestCheckpoint = packet.checkpoints.at(-1);
  const newestHandoff = packet.handoffs.at(-1);
  const required = [
    fixed,
    "",
    "## Latest checkpoint",
    newestCheckpoint ? renderCheckpoint(newestCheckpoint) : "(none)",
    "",
    "## Latest handoff",
    newestHandoff ? renderHandoff(newestHandoff) : "(none)",
  ].join("\n");
  if (required.length > maxChars) {
    // Compact layout keeps the operational handoff (especially next action)
    // ahead of lower-value prose when a pathological task/checkpoint exceeds
    // the entire prompt budget.
    const checkpoint = newestCheckpoint;
    const compact = [
      "# Kanban cold-start context (compact)",
      `Task: ${task.id} [${task.status}] ${clip(task.title, 160)}`,
      `Claim: ${packet.claim?.agentID ?? "unclaimed"}`,
      "",
      "## Latest checkpoint",
      `Next action: ${clip(checkpoint?.nextAction ?? "(none)", 240)}`,
      `Intent: ${clip(checkpoint?.intent ?? "(none)", 180)}`,
      `Summary: ${clip(checkpoint?.summary ?? "(none)", 180)}`,
      `Blockers: ${clip(checkpoint?.blockers.join("; ") || "(none)", 100)}`,
      `Validations: ${clip(checkpoint?.validations.join("; ") || "(none)", 100)}`,
      `Handoff: ${clip(newestHandoff ? `${newestHandoff.id} ${newestHandoff.status} from ${newestHandoff.fromAgent}; next ${newestHandoff.nextAction}` : "(none)", 240)}`,
      "",
      `Dependencies: ${clip(packet.dependencies.map((d) => `${d.id}:${d.status}`).join(", ") || "(none)", 140)}`,
      `Body: ${clip(task.body ?? "(none)", Math.max(80, Math.floor(maxChars / 4)))}`,
      "",
      "[context compacted: full durable history remains in SQLite]",
    ].join("\n");
    return compact.slice(0, maxChars);
  }

  const olderCheckpoints = packet.checkpoints.slice(0, -1).reverse();
  const newestNotes = [...packet.notes].reverse();
  const optional: string[] = [];
  let truncated = packet.truncated;

  for (const checkpoint of olderCheckpoints) {
    const section = `\n\n${renderCheckpoint(checkpoint)}`;
    if (required.length + optional.join("").length + section.length > maxChars) {
      truncated = true;
      break;
    }
    optional.unshift(section);
  }

  let text = [required, "", "## Earlier checkpoints", optional.length ? optional.join("") : "(none)"]
    .join("\n")
    .trimEnd();
  const noteLines: string[] = [];
  for (const note of newestNotes) {
    const line = `\n${renderNote(note)}`;
    const trailer = truncated ? "\n\n[older history omitted]" : "";
    if (text.length + "\n\n## Recent notes".length + noteLines.join("").length + line.length + trailer.length > maxChars) {
      truncated = true;
      break;
    }
    noteLines.unshift(line);
  }
  text += `\n\n## Recent notes\n${noteLines.length ? noteLines.join("").trimStart() : "(none)"}`;
  if (truncated) {
    const marker = "\n\n[older history omitted]";
    if (text.length + marker.length > maxChars) {
      text = text.slice(0, maxChars - marker.length).trimEnd();
    }
    text += marker;
  }
  return text;
}

export function renderTodo(store: KanbanStore): string {
  const name = store.boardName() ?? "Kanban";
  const tasks = store.listTasks();
  const active = tasks.filter((task) => !["done", "cancelled"].includes(task.status));
  const done = tasks.filter((task) => task.status === "done");
  const lines = [
    `# ${name} — generated TODO`,
    "",
    "> Projection only. SQLite is authoritative; do not edit this file as state.",
    "",
    "## Restart here",
  ];

  const inProgress = active.filter((task) => task.status === "in_progress");
  if (!inProgress.length) lines.push("", "No task is currently in progress.");
  for (const task of inProgress) {
    lines.push("", `### ${task.id} — ${task.title}`);
    const checkpoint = store.checkpoints(task.id, 1).at(-1);
    const claim = store.getClaim(task.id);
    lines.push(`- Owner: ${claim?.agentID ?? "unclaimed"}`);
    lines.push(`- Next: ${checkpoint?.nextAction ?? "No checkpoint yet"}`);
    if (checkpoint?.blockers.length) lines.push(`- Blockers: ${checkpoint.blockers.join("; ")}`);
  }

  for (const status of ["blocked", "review", "todo", "backlog"] as const) {
    const group = active.filter((task) => task.status === status);
    lines.push("", `## ${status.replace("_", " ")}`);
    if (!group.length) lines.push("", "(none)");
    else lines.push("", ...group.map(taskLine));
  }
  lines.push("", "## Recently done", "", ...(done.slice(-20).reverse().map(taskLine) || []));
  if (!done.length) lines.push("(none)");
  return `${lines.join("\n")}\n`;
}

File: rust/context.rs
use crate::model::{Checkpoint, ContextPacket, Handoff, RuleSummary, Sitrep, Task, TaskNote};
use crate::store::Store;
use anyhow::{Result, bail};

fn task_line(task: &Task) -> String {
    format!(
        "- {} [{}] P{} {}",
        task.id, task.status, task.priority, task.title
    )
}

fn render_checkpoint(checkpoint: &Checkpoint) -> String {
    let mut lines = vec![
        format!(
            "### Checkpoint {} · {} · {} · {}",
            checkpoint.seq, checkpoint.state, checkpoint.author, checkpoint.created_at
        ),
        format!("Summary: {}", checkpoint.summary),
        format!("Intent: {}", checkpoint.intent),
        format!("Next action: {}", checkpoint.next_action),
    ];
    if !checkpoint.blockers.is_empty() {
        lines.push(format!("Blockers: {}", checkpoint.blockers.join("; ")));
    }
    if !checkpoint.validations.is_empty() {
        lines.push(format!(
            "Validations: {}",
            checkpoint.validations.join("; ")
        ));
    }
    let repository = [
        checkpoint.repo_path.as_deref(),
        checkpoint.branch.as_deref(),
        checkpoint.head_sha.as_deref(),
    ]
    .into_iter()
    .flatten()
    .collect::<Vec<_>>()
    .join(" · ");
    if !repository.is_empty() {
        lines.push(format!("Repository: {repository}"));
    }
    if let Some(dirty) = &checkpoint.dirty_summary {
        lines.push(format!("Working tree: {dirty}"));
    }
    lines.join("\n")
}

/// The most recent sitrep, which for a lane that has been posting is
/// fresher than any checkpoint and cheaper than any handoff.
fn render_sitrep(sitrep: &Sitrep) -> String {
    let mut lines = vec![format!(
        "[{}] {} · {}{}",
        sitrep.lane,
        sitrep.author,
        sitrep.body,
        if sitrep.archived { " (archived)" } else { "" }
    )];
    if let Some(branch) = &sitrep.branch {
        lines.push(format!(
            "branch {branch}{}",
            sitrep
                .head_sha
                .as_deref()
                .map(|sha| format!(" @ {}", &sha[..sha.len().min(8)]))
                .unwrap_or_default()
        ));
    }
    lines.join("\n")
}

fn render_rules(rules: &[RuleSummary]) -> String {
    let globals = rules.iter().filter(|rule| rule.scope == "global").count();
    let projects = rules.len().saturating_sub(globals);
    let mut lines = vec![format!(
        "## Rules ({globals} global · {projects} project; bodies lazy)"
    )];
    if rules.is_empty() {
        lines.push("(none)".to_owned());
        return lines.join("\n");
    }
    for rule in rules {
        let scope = if rule.scope == "global" { "g" } else { "p" };
        let detail = if rule.has_more {
            let global = if rule.scope == "global" {
                " --global"
            } else {
                ""
            };
            format!(
                "  (+{:.1} KB · kb r cat {}{global})",
                rule.bytes as f64 / 1024.0,
                rule.id
            )
        } else {
            String::new()
        };
        lines.push(format!(
            "- [{scope}] {}  {}{detail}",
            rule.id, rule.headline
        ));
    }
    lines.join("\n")
}

fn render_handoff(handoff: &Handoff) -> String {
    let target = handoff
        .to_agent
        .as_deref()
        .unwrap_or("next compatible agent");
    let acceptance = handoff.accepted_by.as_ref().map_or_else(
        || "Accepted by: (pending)".to_owned(),
        |agent| {
            format!(
                "Accepted by: {} · {}",
                agent,
                handoff
                    .accepted_at
                    .map_or_else(|| "-".to_owned(), |at| at.to_string())
            )
        },
    );
    [
        format!(
            "### Handoff {} · {} · {}",
            handoff.id, handoff.status, handoff.reason
        ),
        format!("From: {} · To: {target}", handoff.from_agent),
        format!("Summary: {}", handoff.summary),
        format!("Intent: {}", handoff.intent),
        format!("Next action: {}", handoff.next_action),
        acceptance,
    ]
    .join("\n")
}

fn render_note(note: &TaskNote) -> String {
    format!(
        "- {} · {} · {}: {}",
        note.created_at, note.kind, note.author, note.body
    )
}

fn clip(value: &str, max: usize) -> String {
    if value.chars().count() <= max {
        return value.to_owned();
    }
    let keep = max.saturating_sub(13);
    format!(
        "{}…[truncated]",
        value.chars().take(keep).collect::<String>()
    )
}

fn take_chars(value: &str, max: usize) -> String {
    value.chars().take(max).collect()
}

pub fn render_context(packet: &ContextPacket, max_chars: usize) -> Result<String> {
    if max_chars < 1_000 {
        bail!("max chars must be at least 1000");
    }
    let task = &packet.task;
    let fixed = [
        "# Kanban cold-start context".to_owned(),
        String::new(),
        format!("Generated: {}", packet.generated_at),
        String::new(),
        "## Current task".to_owned(),
        format!("ID: {}", task.id),
        format!("Type: {}", task.task_type),
        format!("Status: {}", task.status),
        format!("Priority: {}", task.priority),
        format!("Title: {}", task.title),
        format!("Body: {}", task.body.as_deref().unwrap_or("(none)")),
        String::new(),
        "## Claim".to_owned(),
        packet.claim.as_ref().map_or_else(
            || "unclaimed".to_owned(),
            |claim| {
                format!(
                    "{} · session {} · expires {}",
                    claim.agent_id,
                    claim.session_id.as_deref().unwrap_or("-"),
                    claim.expires_at
                )
            },
        ),
        String::new(),
        "## Ancestry".to_owned(),
        if packet.ancestors.is_empty() {
            "(none)".to_owned()
        } else {
            packet
                .ancestors
                .iter()
                .map(task_line)
                .collect::<Vec<_>>()
                .join("\n")
        },
        String::new(),
        "## Dependencies".to_owned(),
        if packet.dependencies.is_empty() {
            "(none)".to_owned()
        } else {
            packet
                .dependencies
                .iter()
                .map(task_line)
                .collect::<Vec<_>>()
                .join("\n")
        },
    ]
    .join("\n");
    let newest_checkpoint = packet.checkpoints.last();
    let newest_handoff = packet.handoffs.last();
    let newest_sitrep = packet.sitreps.last();
    let rules = render_rules(&packet.rules);
    let required = format!(
        "{fixed}\n\n{rules}\n\n## Latest sitrep\n{}\n\n## Latest checkpoint\n{}\n\n## Latest handoff\n{}",
        newest_sitrep.map_or_else(|| "(none)".to_owned(), render_sitrep),
        newest_checkpoint.map_or_else(|| "(none)".to_owned(), render_checkpoint),
        newest_handoff.map_or_else(|| "(none)".to_owned(), render_handoff)
    );
    if required.chars().count() > max_chars {
        let checkpoint = newest_checkpoint;
        let compact = [
            "# Kanban cold-start context (compact)".to_owned(),
            format!(
                "Task: {} [{}] {}",
                task.id,
                task.status,
                clip(&task.title, 160)
            ),
            format!(
                "Claim: {}",
                packet
                    .claim
                    .as_ref()
                    .map_or("unclaimed", |claim| claim.agent_id.as_str())
            ),
            rules,
            format!(
                "Sitrep: {}",
                clip(
                    newest_sitrep.map_or("(none)", |value| value.body.as_str()),
                    240
                )
            ),
            String::new(),
            "## Latest checkpoint".to_owned(),
            format!(
                "Next action: {}",
                clip(checkpoint.map_or("(none)", |value| &value.next_action), 240)
            ),
            format!(
                "Intent: {}",
                clip(checkpoint.map_or("(none)", |value| &value.intent), 180)
            ),
            format!(
                "Summary: {}",
                clip(checkpoint.map_or("(none)", |value| &value.summary), 180)
            ),
            format!(
                "Handoff: {}",
                clip(
                    &newest_handoff.map_or_else(
                        || "(none)".to_owned(),
                        |handoff| {
                            format!(
                                "{} {} from {}; next {}",
                                handoff.id, handoff.status, handoff.from_agent, handoff.next_action
                            )
                        }
                    ),
                    240,
                )
            ),
        ]
        .join("\n");
        // Reserved, not appended and hoped for. At the smallest budgets the
        // compact body already overruns, and `take_chars` cut from the end —
        // so the marker was the first thing lost, exactly when the reader most
        // needed telling that the ancestry, dependencies, every earlier
        // checkpoint and every note had gone. Same reservation the full path
        // makes for `[older history omitted]`.
        const MARKER: &str = "\n\n[context compacted: full durable history remains in SQLite]";
        let budget = max_chars.saturating_sub(MARKER.chars().count());
        return Ok(format!(
            "{}{MARKER}",
            take_chars(compact.trim_end(), budget)
        ));
    }

    let mut text = format!("{required}\n\n## Earlier checkpoints");
    let mut truncated = packet.truncated;
    let mut older = Vec::new();
    for checkpoint in packet.checkpoints.iter().rev().skip(1) {
        let section = format!("\n\n{}", render_checkpoint(checkpoint));
        if text.chars().count() + older.join("").chars().count() + section.chars().count()
            > max_chars
        {
            truncated = true;
            break;
        }
        older.insert(0, section);
    }
    if older.is_empty() {
        text.push_str("\n(none)");
    } else {
        text.push_str(&older.join(""));
    }
    text.push_str("\n\n## Recent notes");
    let mut notes = Vec::new();
    for note in packet.notes.iter().rev() {
        let line = format!("\n{}", render_note(note));
        let marker_len = if truncated { 25 } else { 0 };
        if text.chars().count() + notes.join("").chars().count() + line.chars().count() + marker_len
            > max_chars
        {
            truncated = true;
            break;
        }
        notes.insert(0, line);
    }
    if notes.is_empty() {
        text.push_str("\n(none)");
    } else {
        text.push_str(&notes.join(""));
    }
    if truncated {
        let marker = "\n\n[older history omitted]";
        let keep = max_chars.saturating_sub(marker.chars().count());
        if text.chars().count() > keep {
            text = take_chars(text.trim_end(), keep);
        }
        text.push_str(marker);
    }
    Ok(text)
}

pub fn render_todo(store: &Store) -> Result<String> {
    let name = store.board_name()?.unwrap_or_else(|| "Kanban".to_owned());
    let tasks = store.list_tasks(None, None)?;
    let active = tasks
        .iter()
        .filter(|task| task.status != "done" && task.status != "cancelled")
        .collect::<Vec<_>>();
    let done = tasks
        .iter()
        .filter(|task| task.status == "done")
        .collect::<Vec<_>>();
    let mut lines = vec![
        format!("# {name} — generated TODO"),
        String::new(),
        "> Projection only. SQLite is authoritative; do not edit this file as state.".into(),
        String::new(),
        "## Restart here".into(),
    ];
    let in_progress = active
        .iter()
        .filter(|task| task.status == "in_progress")
        .collect::<Vec<_>>();
    if in_progress.is_empty() {
        lines.extend([String::new(), "No task is currently in progress.".into()]);
    }
    for task in in_progress {
        let checkpoint = store.checkpoints(&task.id, 1)?.pop();
        let claim = store.get_claim(&task.id)?;
        lines.extend([
            String::new(),
            format!("### {} — {}", task.id, task.title),
            format!(
                "- Owner: {}",
                claim.as_ref().map_or("unclaimed", |value| &value.agent_id)
            ),
            format!(
                "- Next: {}",
                checkpoint
                    .as_ref()
                    .map_or("No checkpoint yet", |value| &value.next_action)
            ),
        ]);
        if let Some(checkpoint) = checkpoint
            && !checkpoint.blockers.is_empty()
        {
            lines.push(format!("- Blockers: {}", checkpoint.blockers.join("; ")));
        }
    }
    for status in ["blocked", "review", "todo", "backlog"] {
        let group = active
            .iter()
            .filter(|task| task.status == status)
            .collect::<Vec<_>>();
        lines.extend([String::new(), format!("## {}", status.replace('_', " "))]);
        if group.is_empty() {
            lines.extend([String::new(), "(none)".into()]);
        } else {
            lines.push(String::new());
            lines.extend(group.into_iter().map(|task| task_line(task)));
        }
    }
    lines.extend([String::new(), "## Recently done".into(), String::new()]);
    if done.is_empty() {
        lines.push("(none)".into());
    } else {
        lines.extend(done.into_iter().rev().take(20).map(task_line));
    }
    Ok(format!("{}\n", lines.join("\n")))
}
